### PR TITLE
JAVA-3369: update FieldNameValidator to detect empty update documents

### PR DESCRIPTION
**Pull request by jstewart-mongo**
Evergreen patch: https://evergreen.mongodb.com/version/5d5acee63066153417a64ebc

Changes to detect when an update document passed to `findOneAndUpdate` is empty.

### CODE REVIEW

File: bson/src/main/org/bson/FieldNameValidator.java [hunk 1]
```diff
@@ -37,4 +37,18 @@
      * @return a non-null validator
      */
     FieldNameValidator getValidatorForField(String fieldName);
+
+    /**
+     * Start validation of a single document.
+     *
+     * @since 4.0
+     */
+    void start();
```
Comment (rozza) [inline bson/src/main/org/bson/FieldNameValidator.java]: As we are now in Java 8 - would it be better to use a default no-op method for start and end?

Comment (jyemin) [reply bson/src/main/org/bson/FieldNameValidator.java]: +1

The advantage is that existing validators (our own and third-party) will continue to work without change, and only have to override the default methods if they really need to.

File: driver-core/src/main/com/mongodb/internal/validator/MappedFieldNameValidator.java [hunk 2]
```diff
@@ -56,4 +56,14 @@ public FieldNameValidator getValidatorForField(final String fieldName) {
             return defaultValidator;
         }
     }
+
+    @Override
+    public void start() {
+        defaultValidator.start();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/validator/MappedFieldNameValidator.java]: I'm not sure these overrides are necessary, or correct.  Are there tests that prove it either way?

File: bson/src/main/org/bson/FieldNameValidator.java (hunk 1 above)
Comment (jyemin) [reply bson/src/main/org/bson/FieldNameValidator.java]: @rozza resolving this as you've already LGTM'd

File: driver-core/src/main/com/mongodb/internal/validator/MappedFieldNameValidator.java (hunk 2 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/validator/MappedFieldNameValidator.java]: Let me know on this one.

Comment (jstewart-mongo) [reply driver-core/src/main/com/mongodb/internal/validator/MappedFieldNameValidator.java]: I have debugged this, and there are no tests that prove it either way. My thinking here was to emulate how `defaultValidator` was used in other methods, such as `validate()`. I have removed the `start()` and `end()` methods from this class.

**Pull request closed without merging**